// .ignored/framer-motion/dist/cjs/dom-entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var indexLegacy = require('./index-legacy-4f1fdbbf.js');



exports.MotionValue = indexLegacy.MotionValue;
exports.animate = indexLegacy.animate;
exports.anticipate = indexLegacy.anticipate;
exports.backIn = indexLegacy.backIn;
exports.backInOut = indexLegacy.backInOut;
exports.backOut = indexLegacy.backOut;
exports.cancelFrame = indexLegacy.cancelFrame;
exports.cancelSync = indexLegacy.cancelSync;
exports.circIn = indexLegacy.circIn;
exports.circInOut = indexLegacy.circInOut;
exports.circOut = indexLegacy.circOut;
exports.clamp = indexLegacy.clamp;
exports.createScopedAnimate = indexLegacy.createScopedAnimate;
exports.cubicBezier = indexLegacy.cubicBezier;
exports.delay = indexLegacy.delay;
exports.distance = indexLegacy.distance;
exports.distance2D = indexLegacy.distance2D;
exports.easeIn = indexLegacy.easeIn;
exports.easeInOut = indexLegacy.easeInOut;
exports.easeOut = indexLegacy.easeOut;
exports.frame = indexLegacy.frame;
exports.frameData = indexLegacy.frameData;
exports.inView = indexLegacy.inView;
exports.interpolate = indexLegacy.interpolate;
Object.defineProperty(exports, 'invariant', {
	enumerable: true,
	get: function () { return indexLegacy.invariant; }
});
exports.mirrorEasing = indexLegacy.mirrorEasing;
exports.mix = indexLegacy.mix;
exports.motionValue = indexLegacy.motionValue;
exports.pipe = indexLegacy.pipe;
exports.progress = indexLegacy.progress;
exports.reverseEasing = indexLegacy.reverseEasing;
exports.scroll = indexLegacy.scroll;
exports.stagger = indexLegacy.stagger;
exports.steps = indexLegacy.steps;
exports.stepsOrder = indexLegacy.stepsOrder;
exports.sync = indexLegacy.sync;
exports.transform = indexLegacy.transform;
Object.defineProperty(exports, 'warning', {
	enumerable: true,
	get: function () { return indexLegacy.warning; }
});
exports.wrap = indexLegacy.wrap;
