// eslint-plugin-import/lib/rules/no-amd.js
'use strict';




var _docsUrl = require('../docsUrl');var _docsUrl2 = _interopRequireDefault(_docsUrl);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { 'default': obj };}

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    type: 'suggestion',
    docs: {
      category: 'Module systems',
      description: 'Forbid AMD `require` and `define` calls.',
      url: (0, _docsUrl2['default'])('no-amd') },

    schema: [] },


  create: function () {function create(context) {
      return {
        'CallExpression': function () {function CallExpression(node) {
            if (context.getScope().type !== 'module') return;

            if (node.callee.type !== 'Identifier') return;
            if (node.callee.name !== 'require' &&
            node.callee.name !== 'define') return;

            // todo: capture define((require, module, exports) => {}) form?
            if (node.arguments.length !== 2) return;

            var modules = node.arguments[0];
            if (modules.type !== 'ArrayExpression') return;

            // todo: check second arg type? (identifier or callback)

            context.report(node, 'Expected imports instead of AMD ' + String(node.callee.name) + '().');
          }return CallExpression;}() };


    }return create;}() }; /**
                           * @fileoverview Rule to prefer imports to AMD
                           * @author Jamund Ferguson
                           */
