// .ignored/framer-motion/dist/es/dom-entry.mjs
export { MotionValue, motionValue } from './value/index.mjs';
export { animate, createScopedAnimate } from './animation/animate.mjs';
export { scroll } from './render/dom/scroll/index.mjs';
export { inView } from './render/dom/viewport/index.mjs';
export { anticipate } from './easing/anticipate.mjs';
export { backIn, backInOut, backOut } from './easing/back.mjs';
export { circIn, circInOut, circOut } from './easing/circ.mjs';
export { easeIn, easeInOut, easeOut } from './easing/ease.mjs';
export { cubicBezier } from './easing/cubic-bezier.mjs';
export { mirrorEasing } from './easing/modifiers/mirror.mjs';
export { reverseEasing } from './easing/modifiers/reverse.mjs';
export { stagger } from './animation/utils/stagger.mjs';
export { transform } from './utils/transform.mjs';
export { clamp } from './utils/clamp.mjs';
export { delay } from './utils/delay.mjs';
export { distance, distance2D } from './utils/distance.mjs';
export { invariant, warning } from './utils/errors.mjs';
export { interpolate } from './utils/interpolate.mjs';
export { mix } from './utils/mix.mjs';
export { pipe } from './utils/pipe.mjs';
export { progress } from './utils/progress.mjs';
export { wrap } from './utils/wrap.mjs';
export { cancelFrame, frame, steps, stepsOrder } from './frameloop/index.mjs';
export { frameData } from './frameloop/data.mjs';
export { cancelSync, sync } from './frameloop/index-legacy.mjs';
