// .ignored/next/dist/server/app-render/entry-base.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
0 && (module.exports = {
    AppRouter: null,
    LayoutRouter: null,
    RenderFromTemplateContext: null,
    staticGenerationAsyncStorage: null,
    requestAsyncStorage: null,
    actionAsyncStorage: null,
    staticGenerationBailout: null,
    createSearchParamsBailoutProxy: null,
    serverHooks: null,
    renderToReadableStream: null,
    decodeReply: null,
    decodeAction: null,
    preloadStyle: null,
    preloadFont: null,
    preconnect: null,
    StaticGenerationSearchParamsBailoutProvider: null
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    AppRouter: function() {
        return AppRouter;
    },
    LayoutRouter: function() {
        return LayoutRouter;
    },
    RenderFromTemplateContext: function() {
        return RenderFromTemplateContext;
    },
    staticGenerationAsyncStorage: function() {
        return staticGenerationAsyncStorage;
    },
    requestAsyncStorage: function() {
        return requestAsyncStorage;
    },
    actionAsyncStorage: function() {
        return actionAsyncStorage;
    },
    staticGenerationBailout: function() {
        return staticGenerationBailout;
    },
    createSearchParamsBailoutProxy: function() {
        return createSearchParamsBailoutProxy;
    },
    serverHooks: function() {
        return serverHooks;
    },
    renderToReadableStream: function() {
        return renderToReadableStream;
    },
    decodeReply: function() {
        return decodeReply;
    },
    decodeAction: function() {
        return decodeAction;
    },
    preloadStyle: function() {
        return preloadStyle;
    },
    preloadFont: function() {
        return preloadFont;
    },
    preconnect: function() {
        return preconnect;
    },
    StaticGenerationSearchParamsBailoutProvider: function() {
        return StaticGenerationSearchParamsBailoutProvider;
    }
});
const { default: AppRouter  } = require("next/dist/client/components/app-router");
const { default: LayoutRouter  } = require("next/dist/client/components/layout-router");
const { default: RenderFromTemplateContext  } = require("next/dist/client/components/render-from-template-context");
const { staticGenerationAsyncStorage  } = require("next/dist/client/components/static-generation-async-storage");
const { requestAsyncStorage  } = require("next/dist/client/components/request-async-storage");
const { actionAsyncStorage  } = require("next/dist/client/components/action-async-storage");
const { staticGenerationBailout  } = require("next/dist/client/components/static-generation-bailout");
const { default: StaticGenerationSearchParamsBailoutProvider  } = require("next/dist/client/components/static-generation-searchparams-bailout-provider");
const { createSearchParamsBailoutProxy  } = require("next/dist/client/components/searchparams-bailout-proxy");
const serverHooks = require("next/dist/client/components/hooks-server-context");
const { renderToReadableStream , decodeReply , decodeAction  } = require("react-server-dom-webpack/server.edge");
const { preloadStyle , preloadFont , preconnect  } = require("next/dist/server/app-render/rsc/preloads");
